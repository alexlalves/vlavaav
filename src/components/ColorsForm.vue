<template>
  <form @submit.prevent="formSubmit">
    <p v-if="errorMessageToShow">
      {{ errorMessageToShow }}
    </p>

    <textarea
      class="colors-form__input-area"
      :placeholder=samplePalette
      v-model="pastedInfo"
    />

    <button
      class="colors-form__submit-button"
      type="submit"
    >
      Generate
    </button>
  </form>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

@Component
export default class ColorsForm extends Vue {
  private samplePalette = [
    'Paste here or just click \'Generate\'',
    'and see these colors in action:',
    '',
    'V: #ff0000,',
    'L: #ff8800,',
    'A: #ffff00,',
    'V: #00ff00,',
    'A: #0000ff,',
    'A: #8800ff,',
    'V: #ff00ff.',
  ].join('\n');

  private errorMessageToShow = '';

  private pastedInfo = '';

  public formSubmit() {
    const colors = this.findColors(this.pastedInfo || this.samplePalette);

    if (colors.length) {
      this.$router.push({ name: 'Palette', params: { colors } });
    } else {
      this.errorMessageToShow = 'No colors of the #RRGGBB format were found in the submitted text.';
    }
  }

  public findColors(pastedText: string): string {
    const colors = pastedText.match(/(?:#)[0-9a-fA-F]{6}/g) || [];

    return [...new Set(
      colors.map(color => color.toLowerCase().replace('#', '')),
    )].join(',');
  }
}
</script>

<style scoped lang="scss">
.colors-form__input-area {
  background-color: #111111;
  box-sizing: border-box;
  color: #ffffff;
  display: block;
  font-size: 1.25em;
  line-height: 1.25em;
  margin: 0;
  margin-bottom: 32px;
  min-height: 280px;
  padding: 1em 1em;
  resize: vertical;
  width: 100%;

  @media (min-width: 480px) {
    min-height: 320px;
  }
}

.colors-form__submit-button {
  background-color: #606060;
  border: none;
  box-sizing: border-box;
  color: #ffffff;
  cursor: pointer;
  font-size: 1.1em;
  margin-bottom: 0;
  padding: 1em;
  transition: 0.2s ease-in-out;
  width: 100%;

  &:hover {
    background-color: #7a7a7a;
  }
}
</style>
