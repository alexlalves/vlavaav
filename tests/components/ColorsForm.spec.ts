import { shallowMount } from '@vue/test-utils';
import ColorsForm from '@/components/ColorsForm.vue';

const push = jest.fn();

describe('ColorsForm', () => {
  def('wrapper', () => shallowMount(ColorsForm, {
    mocks: {
      $router: {
        push,
      },
    },
  }));

  afterEach(() => {
    push.mockClear();
  });

  it('matches snapshot', () => {
    expect(get('wrapper').element).toMatchSnapshot();
  });

  describe('findColors', () => {
    describe('when text with no color is passed', () => {
      it('returns an empty list', () => {
        expect(
          get('wrapper').vm.findColors('Color: white :('),
        ).toEqual('');
      });
    });

    describe('when an uppercase color is passed', () => {
      it('returns that color', () => {
        expect(
          get('wrapper').vm.findColors('Color: #FF00FF'),
        ).toMatch('ff00ff');
      });
    });

    describe('when more than one color is passed', () => {
      describe('when they are unique', () => {
        it('returns those colors', () => {
          const colors = get('wrapper').vm.findColors('ColorA: #FF00FF #abcde0');

          expect(colors).toMatch('ff00ff');
          expect(colors).toMatch('abcde0');
        });
      });

      describe('when they are duplicated', () => {
        it('returns only one of the duplicated color', () => {
          const colors = get('wrapper').vm.findColors('ColorA: #FF00FF #ff00ff');
          expect(colors).toEqual('ff00ff');
        });
      });
    });
  });

  describe('when submitting colors', () => {
    beforeEach(() => {
      get('wrapper').setData({
        pastedInfo: get('userInput'),
        samplePalette: '#ff00ff',
      });
      get('wrapper').vm.formSubmit();
    });

    describe('when no user input has been provided', () => {
      def('userInput', () => '');

      it('accesses the Palette view with the default colors', () => {
        expect(push).toBeCalledWith({
          name: 'Palette', params: { colors: 'ff00ff' },
        });
      });
    });

    describe('when user input has no colors provided', () => {
      def('userInput', () => 'Just a normal sentence');

      it('shows an error message', () => {
        expect(get('wrapper').vm.$data.pastedInfo).not.toBeFalsy();
        expect(push).not.toBeCalled();
      });
    });

    describe('when user input provides colors', () => {
      def('userInput', () => 'A color: #abcdef');

      it('accesses the Palette view with the provided colors', () => {
        expect(push).toBeCalledWith({
          name: 'Palette', params: { colors: 'abcdef' },
        });
      });
    });
  });
});
