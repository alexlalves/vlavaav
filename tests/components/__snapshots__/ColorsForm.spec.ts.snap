// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorsForm matches snapshot 1`] = `
<form>
  <!---->
   
  <textarea
    class="colors-form__input-area"
    placeholder="Paste here or just click 'Generate'
and see these colors in action:

V: #ff0000,
L: #ff8800,
A: #ffff00,
V: #00ff00,
A: #0000ff,
A: #8800ff,
V: #ff00ff."
  />
   
  <button
    class="colors-form__submit-button"
    type="submit"
  >
    
    Generate
  
  </button>
</form>
`;
